flexible-event/flexible_event_privacy.py

```diff
@@ -0,0 +1,164 @@
+    """Returns the maximum information for a source using the flexible event API."""
+
+    flip_prob = flip_probability_dp(num_states, epsilon)
+    return capacity_q_ary_symmetric_channel(log(num_states, 2),
+                                            flip_prob*(num_states-1)/num_states)
+
+
+def get_config(json: dict) -> ApiConfig:
+    max_event_level_reports = json.get('max_event_level_reports', 20)
+    per_trigger_data_configs = []
+    for spec in json['trigger_specs']:
+        num_data_types = len(spec['trigger_data'])
+        num_windows = len(spec['event_report_windows']['end_times'])
+        num_buckets = len(spec['summary_buckets']
+                          ) if 'summary_buckets' in spec else 20
+        per_trigger_data_configs.extend(
+            [(num_windows, num_buckets)] * num_data_types)
+
+    return ApiConfig(max_event_level_reports, per_trigger_data_configs)
+
+
+if __name__ == "__main__":
+    DESCRIPTION = '''\
+  flexible_event_privacy.py is a utility to ingest configurations for the flexible
+  event-level reports in the Attribution Reporting API. It reads a JSON from stdin
+  matching the input in the request header Attribution-Reporting-Register-Source.
+
+  It optionally also accepts windows and per-trigger-data summary buckets from command
+  line arguments.
+  '''
+
+    parser = argparse.ArgumentParser(
+        description=DESCRIPTION, formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument('-m', '--max_event_level_reports',
+                        type=int, default=20)
+    parser.add_argument('-e', '--epsilon', type=float, default=14)
+
+    def comma_separated_ints(string):
+        return [int(i) for i in string.split(',')]
+
+    WINDOW_HELP = '''\
+  Comma separated integers representing the number of reporting windows
+  for each possible trigger_data.
+  '''
+    parser.add_argument('-w', '--windows',
+                        type=comma_separated_ints, help=WINDOW_HELP)
+
+    BUCKETS_HELP = '''\
+  Comma separated integers representing the maximum number of summary buckets
+  for each possible trigger_data.
+  '''
+    parser.add_argument('-b', '--buckets',
+                        type=comma_separated_ints, help=BUCKETS_HELP)
+    args = parser.parse_args()
+
+    api_config: ApiConfig = None
+    if args.windows and args.buckets:
+        assert len(args.windows) == len(args.buckets)
+        per_trigger_configs = list(zip(args.windows, args.buckets))
+        api_config = ApiConfig(
+            args.max_event_level_reports, per_trigger_configs)
+    else:
+        api_config = get_config(json.load(sys.stdin))
+
+    num_states = num_flexible_states(api_config)
+    info_gain = max_information_gain(num_states, args.epsilon)
+    flip_prob = flip_probability_dp(num_states, args.epsilon)
+
+    print(f"Num states: {num_states}")
+    print(f"Information gain: {info_gain:.2f} bits")
+    print(f"Flip percent: {100 * flip_prob:.5f}%")
```

Comment:
Please add a newline at the end of the file.

Reply:
Done